os/kernel/binary_manager/binary_manager.c

```diff
@@ -0,0 +1,407 @@
+#include <mqueue.h>
+#include <errno.h>
+#include <debug.h>
+#include <fcntl.h>
+#include <string.h>
+#if !defined(CONFIG_DISABLE_SIGNALS)
+#include <signal.h>
+#endif
+#ifdef CONFIG_BOARD_ASSERT_AUTORESET
+#include <sys/boardctl.h>
+#include <stdlib.h>
+#endif
+#include <tinyara/sched.h>
+#include <tinyara/binary_manager.h>
+
+#include "group/group.h"
+#include "binary_manager.h"
+
+/****************************************************************************
+ * Private Definitions
+ ****************************************************************************/
+#define BINMGR_DEVNAME_FMT "/dev/mtdblock%d"
+#define BINMGR_DEVNAME_LEN 16
+
+struct binary_header_s {
+	uint16_t header_size;
+	uint16_t bin_type;
+	uint32_t bin_size;
+	char bin_name[BIN_NAME_MAX];
+	char bin_ver[BIN_VER_MAX];
+	char kernel_ver[KERNEL_VER_MAX];
+	uint32_t jump_addr;
+};
+typedef struct binary_header_s binary_header_t;
+
+struct binary_config_s {
+	char name[BIN_NAME_MAX];
+	uint16_t use_flag;
+};
+typedef struct binary_config_s binary_config_t;
+
+struct binary_info_s {
+	pid_t binid;
+	uint32_t bin_size;
+	uint16_t bin_offset;
+	uint16_t inuse_idx;
+	uint32_t part_size;
+	uint16_t part_num[2];
+	char name[BIN_NAME_MAX];
+	char bin_ver[BIN_VER_MAX];
+	char kernel_ver[KERNEL_VER_MAX];
+};
+typedef struct binary_info_s binary_info_t;
+
+binary_info_t bin_table[USER_BIN_COUNT];
+int g_bin_count;
+int g_binconfig_part;
+
+#define BIN_ID(bin_idx)					bin_table[bin_idx].binid
+#define BIN_SIZE(bin_idx)				bin_table[bin_idx].bin_size
+#define BIN_OFFSET(bin_idx)				bin_table[bin_idx].bin_offset
+#define BIN_USEIDX(bin_idx)				bin_table[bin_idx].inuse_idx
+
+#define BIN_PARTSIZE(bin_idx)				bin_table[bin_idx].part_size
+#define BIN_PARTNUM(bin_idx, part_idx)			bin_table[bin_idx].part_num[part_idx]
+
+#define BIN_NAME(bin_idx)				bin_table[bin_idx].name
+#define BIN_VER(bin_idx)				bin_table[bin_idx].bin_ver
+#define BIN_KERNEL_VER(bin_idx)				bin_table[bin_idx].kernel_ver
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+void binary_manager_register_bin_part(int part_num, char *name, int partsize)
+{
+	int idx;
+
+	if (part_num < 0 || name == NULL || partsize <= 0) {
+		bmdbg("ERROR: Invalid input \n");
+		return;
+	}
+
+	for (idx = 0; idx < g_bin_count; idx++) {
+		if (!strncmp(BIN_NAME(idx), name, strlen(name) + 1)) {
+			BIN_PARTNUM(idx, 1) = part_num;
+			bmvdbg("[Second] %s %d %d \n", BIN_NAME(idx), BIN_PARTNUM(idx, 0), BIN_PARTNUM(idx, 1));
+			return;
+		}
+	}
+
+	/* Register new partition */
+	BIN_PARTNUM(g_bin_count, 0) = part_num;
+	BIN_PARTSIZE(g_bin_count) = partsize;
+	strncpy(BIN_NAME(g_bin_count), name, BIN_NAME_MAX);
+
+	bmvdbg("[NEW %d] %s size %d %d \n", g_bin_count, BIN_NAME(g_bin_count), BIN_PARTSIZE(g_bin_count), BIN_PARTNUM(g_bin_count, 0));
+	g_bin_count++;
+}
+
+void binary_manager_register_loadparam_part(int part_num)
```

Comment:
What's the purpose of `loadparam`?

Reply:
It would be used to save values like state.